### ISSUE TITLE
Random exception with `Optional` parameter

### ISSUE THREAD

**Issue by azmeuk** [felix-martel/pydanclick#19]
With pydantic 2.9.0 and python 3.12, I randomly get an exception with the following snippet:

```python
import click
from pydanclick import from_pydantic
from pydantic import BaseModel
from typing import Optional


class Bar(BaseModel):
    baz: str = "baz"


class Foo(BaseModel):
    bar: Optional[Bar] = None


@click.command()
@from_pydantic(Foo)
def cli(foo: Foo):
    ...


if __name__ == "__main__":
    cli()
```

```python
python foobar.py
Traceback (most recent call last):
  File "/home/eloi/dev/yaal/scim2-cli/foobar.py", line 22, in <module>
    cli()
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/pydanclick/main.py", line 66, in wrapped
    kwargs[variable_name] = validator(kwargs)
                            ^^^^^^^^^^^^^^^^^
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/pydanclick/model/validation.py", line 45, in model_validate_kwargs
    raw_model = _unflatten_dict(flat_model)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/eloi/.cache/pypoetry/virtualenvs/scim2-cli-KsA89t9a-py3.12/lib/python3.12/site-packages/pydanclick/model/validation.py", line 67, in _unflatten_dict
    current_dict[parts[-1]] = v
    ~~~~~~~~~~~~^^^^^^^^^^^
TypeError: 'NoneType' object does not support item assignment
```

When using the `| None` syntax instead of `Optional`, the snippet never crashes, though I am not sure why:

```python
class Foo(BaseModel):
    bar: Bar | None = None
```

I am not encountering the exception if the type of bar is `str` for instance:

```python
class Foo(BaseModel):
    bar: Optional[str] = None
```

**Comment by azmeuk** [comment]
`flat_model` has different values depending on the situations:
https://github.com/felix-martel/pydanclick/blob/2468b62ef0ac9ccd93dedec0eb3d612b537b8f4f/pydanclick/model/validation.py#L44

On cases the exception is raised, it has the value of `{'bar': None, 'bar.baz': 'baz'}`, on cases the exception is not raised it has the value of `{'bar.baz': 'baz', 'bar': None}`.

**Comment by felix-martel** [comment]
Thanks for the bug report, I'll look into it!

**Comment by felix-martel** [comment]
Can reproduce. It was caused by the way we treated default value and was solved by #17 
I'll create a new release right away

**Comment by azmeuk** [comment]
I confirm this is solved with 0.3. Thank you!
